File: step-01-basics/sieve_of_eratosthenes/sol.py
"""
problem :
link    :
pattern :
brute   : O( ) time, O( ) space
optimal : O( ) time, O( ) space
"""
from typing import List, Optional
from collections import defaultdict, deque, Counter
from heapq import heappush, heappop, heapify
from bisect import bisect_left, bisect_right

def sieve(n):
    isPrime = [True]*(n+1)
    isPrime[0] = isPrime[1] = False
    for i in range(2,n+1):
        if(isPrime[i]):
            for j in range(i*i,n+1,i):
                isPrime[j]=False
    for i in range(n+1):
        if(isPrime[i]):
            print(f"{i} ",end="")

def solve():
    sieve(25)


if __name__ == "__main__":
    solve()

File: step-01-basics/sieve_of_eratosthenes/sol.cpp
// problem :
// link    :
// pattern :
// brute   : O( ) time, O( ) space
// better  : O( ) time, O( ) space
// optimal : O( ) time, O( ) space

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <queue>
#include <stack>
#include <climits>
#include <cmath>
using namespace std;

vector<bool> sieveOfEratosthenes(int n)
{
    vector<bool> isPrime(n + 1, true);
    isPrime[0] = isPrime[1] = false;
    for (int i = 2; i * i < n; i++)
    {
        if (isPrime[i])
        {
            for (int j = i * i; j <= n; j += i)
            {
                isPrime[j] = false;
            }
        }
    }
    return isPrime;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n = 25;
    vector<bool> isPrime = sieveOfEratosthenes(n);
    for (int i = 0; i <= n; i++)
    {
        if (isPrime[i])
        {
            cout << i << " ";
        }
    }

    return 0;
}
